gnome-initial-setup/pages/site/gis-site-page.c

```diff
@@ -0,0 +1,473 @@
+          json_reader_end_member (reader);
+
+          json_reader_read_member (reader, "region");
+          region = json_reader_get_string_value (reader);
+          json_reader_end_member (reader);
+
+          json_reader_read_member (reader, "country");
+          country = json_reader_get_string_value (reader);
+          json_reader_end_member (reader);
+
+          site = gis_site_new (id, facility, street, locality, region, country);
+          g_ptr_array_add (sites, site);
+        }
+      else
+        {
+          sites_usable = FALSE;
+          goto out;
+        }
+      json_reader_end_element (reader);
+    }
+
+out:
+  g_clear_object (&reader);
+  g_clear_object (&parser);
+  if (!sites_usable)
+    g_clear_pointer (&sites, g_ptr_array_unref);
+
+  return sites;
+}
+
+static void
+gis_site_page_constructed (GObject *object)
+{
+  GisSitePage *page = GIS_SITE_PAGE (object);
+  GisSitePagePrivate *priv = gis_site_page_get_instance_private (page);
+  gboolean visible = TRUE;
+
+  G_OBJECT_CLASS (gis_site_page_parent_class)->constructed (object);
+
+  if (!g_file_test (GIS_SITE_PAGE_SITES_FILE, G_FILE_TEST_EXISTS))
+    {
+      visible = FALSE;
+      goto out;
+    }
+
+  /* make all the fields insensitive to start */
+  manual_check_toggled (GTK_TOGGLE_BUTTON (priv->manual_check), page);
+
+  // let the ID field show its current value (if populated based on a matched
+  // pre-defined site) but don't let the user set it when editing manually
+  gtk_widget_set_can_focus (priv->id_entry, FALSE);
+  gtk_widget_set_sensitive (priv->id_entry, FALSE);
+
+  g_signal_connect (priv->search_entry,
+                    "notify::" GIS_SITE_SEARCH_ENTRY_PROP_SITE,
+                    G_CALLBACK (entry_site_changed), page);
+
+  g_signal_connect (priv->manual_check, "toggled",
+                    G_CALLBACK (manual_check_toggled), page);
+
+  g_signal_connect (priv->facility_entry,
+                    "notify::text-length",
+                    G_CALLBACK (update_page_validation), page);
+
+  g_signal_connect (priv->street_entry,
+                    "notify::text-length",
+                    G_CALLBACK (update_page_validation), page);
+
+  g_signal_connect (priv->locality_entry,
+                    "notify::text-length",
+                    G_CALLBACK (update_page_validation), page);
+
+  g_signal_connect (priv->region_entry,
+                    "notify::text-length",
+                    G_CALLBACK (update_page_validation), page);
+
+  g_signal_connect (priv->country_entry,
+                    "notify::text-length",
+                    G_CALLBACK (update_page_validation), page);
+
+out:
+  gtk_widget_set_visible (GTK_WIDGET (page), visible);
+}
+
+static void
+gis_site_page_class_init (GisSitePageClass *klass)
+{
+  GisPageClass *page_class = GIS_PAGE_CLASS (klass);
+  GObjectClass *object_class = G_OBJECT_CLASS (klass);
+
+  gtk_widget_class_set_template_from_resource (GTK_WIDGET_CLASS (klass),
```

Comment:
You could cast to `GtkWidgetClass` at the beginning of the function to avoid doing it so many times here.